File: src/main/java/com/aau/wizard/model/Card.java
package com.aau.wizard.model;

import com.aau.wizard.model.enums.CardType;
import com.aau.wizard.model.enums.CardSuit;

import java.util.Objects;

// fixme extract Special/jester and special/wizard into classes for polymorphism
public class Card {
    private final CardSuit cardSuit;
    private final int value;
    private final CardType type;

    public Card(CardSuit cardSuit, int value) {
        this.cardSuit = cardSuit;
        this.value = value;
        if (value == 0) {
            this.type = CardType.JESTER;
        } else if (value == 14) {
            this.type = CardType.WIZARD;
        } else {
            this.type = CardType.NUMBER;
        }
    }

    public CardSuit getSuit() {
        return cardSuit;
    }

    public int getValue() {
        return value;
    }

    public CardType getType() {
        return type;
    }

    @Override
    public String toString() {
        switch (type) {
            case WIZARD:
                return "Wizard (" + cardSuit.name().toLowerCase() + ")";
            case JESTER:
                return "Jester (" + cardSuit.name().toLowerCase() + ")";
            default:
                return value + " of " + cardSuit.name().toLowerCase();
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Card)) return false;
        Card card = (Card) o;
        return value == card.value &&
                cardSuit == card.cardSuit &&
                type == card.type;
    }

    @Override
    public int hashCode() {
        return Objects.hash(cardSuit, value, type);
    }
}

File: src/main/kotlin/com/aau/wizard/core/cards/Card.kt
package com.aau.wizard.core.cards

enum class Suit {
    RED, YELLOW, BLUE, GREEN, SPECIAL
}

enum class CardType {
    NUMBER, WIZARD, JESTER
}

data class Card(
    val suit: Suit,
    val value: Int, // 1- 13 für Zahlenkarten, 0 für den Jester, 14 fürn Wizard
    val type: CardType = when {
        value == 0 -> CardType.JESTER
        value == 14 -> CardType.WIZARD
        else -> CardType.NUMBER
    }
) {
    override fun toString(): String = when (type) {
        CardType.WIZARD -> "Wizard (${suit.name.lowercase()})"
        CardType.JESTER -> "Jester (${suit.name.lowercase()})"
        else -> "$value of ${suit.name.lowercase()}"
    }
}